From 450a50eda06a30101e972d1a6b285c39a86477dd Mon Sep 17 00:00:00 2001
From: William Strecker-Kellogg <willsk@bnl.gov>
Date: Fri, 25 Oct 2013 13:18:37 -0400
Subject: [PATCH] New parameter for setting user execdir permissions

User can set JOB_EXECDIR_PERMISSIONS variable to one of:
    * user (default)
    * group
    * world
that will make the job's scratch directory have permissions 0700, 0770,
or 0777 respectively (subject to  & ~umask).

At the RACF the PHENIX experiment relied on the the directory being
at least group-readable, since the setuid copy-as-user program for
transferring output on behalf of another user needs to read files in
this directory.
---
 doc/admin-man/configure.tex             | 11 +++++++++++
 src/condor_starter.V6.1/baseStarter.cpp | 15 +++++++++++++++
 src/condor_utils/param_info.in          | 16 ++++++++++++++++
 3 files changed, 42 insertions(+)

diff --git a/doc/admin-man/configure.tex b/doc/admin-man/configure.tex
index c6487ba..5bfb650 100644
--- a/doc/admin-man/configure.tex
+++ b/doc/admin-man/configure.tex
@@ -6512,6 +6512,17 @@ REM Run the actual job now
   than the default \Expr{Users} for the \Expr{condor-reuse-slot<X>} run account.
   Do \emph{not} place the local group name within quotation marks.
 
+\label{param:JobExecdirPermissions}
+\item[\Macro{JOB\_EXECDIR\_PERMISSIONS}]
+  Control the permissions on the job's scratch directory. Defaults to
+  \Expr{user} which means permissions of \emph{0700}. Possible values are
+  \Expr{user}, \Expr{group}, and \Expr{world}. If set to \Expr{group} then the
+  directory is group-accessible, permissions of \emph{0770}; likewise if set
+  to \Expr{world} the directory is created with permissions \emph{0777}. All
+  of these are subject to umask, which defaults to 022, meaning setting this
+  to \Expr{world} will \emph{not} typically result in a world-writable
+  directory being created.
+
 
 \end{description}
 
diff --git a/src/condor_starter.V6.1/baseStarter.cpp b/src/condor_starter.V6.1/baseStarter.cpp
index a40c4fa..fa3842e 100644
--- a/src/condor_starter.V6.1/baseStarter.cpp
+++ b/src/condor_starter.V6.1/baseStarter.cpp
@@ -1800,7 +1800,22 @@ CStarter::createTempExecuteDir( void )
 		// might be using glexec.  glexec relies on being able to read the
 		// contents of the execute directory as a non-condor user, so in that
 		// case, use 0755.  for all other cases, use the more-restrictive 0700.
+
 		int dir_perms = 0700;
+
+		// Parameter JOB_EXECDIR_PERMISSIONS can be user / group / world and
+		// defines permissions on execute directory (subject to umask)
+		char *who = param("JOB_EXECDIR_PERMISSIONS");
+		if(who != NULL)	{
+			if(!strcasecmp(who, "user"))
+				dir_perms = 0700;
+			else if(!strcasecmp(who, "group"))
+				dir_perms = 0770;
+			else if(!strcasecmp(who, "world"))
+				dir_perms = 0777;
+			free(who);
+		}
+
 #if defined(LINUX)
 		if(glexecPrivSepHelper()) {
 			dir_perms = 0755;
diff --git a/src/condor_utils/param_info.in b/src/condor_utils/param_info.in
index b2c7c23..ac709eb 100644
--- a/src/condor_utils/param_info.in
+++ b/src/condor_utils/param_info.in
@@ -10970,3 +10970,19 @@ usage=
 url=http://cs.wisc.edu/condor/manual/v7.3/3_3Configuration.html#SECTION
 tags=starter
 id=615
+
+[JOB_EXECDIR_PERMISSIONS]
+aliases=
+default=user
+version=8.0.5
+range=.*
+state=default
+type=string
+is_macro=false
+reconfig=true
+customization=seldom
+friendly_name=Permissions on the job's execute directory
+usage=
+url=http://cs.wisc.edu/condor/manual/v8.0/3_3Configuration.html#SECTION
+tags=starter
+id=616
-- 
1.8.3.1

